codegen/src/main.zig

```diff
@@ -634,7 +752,465 @@ fn generateMetadataFunction(operation_name: []const u8, state: GenerationState,
     try outputIndent(child_state, writer);
     _ = try writer.write("}\n");
     try outputIndent(state, writer);
-    try writer.writeByte('}');
+
+    if (options.end_structure) {
+        try writer.writeByte('}');
+    }
+}
+
+fn findTrait(trait_type: smithy.TraitType, traits: []smithy.Trait) ?smithy.Trait {
+    for (traits) |trait| {
+        if (trait == trait_type) {
+            return trait;
+        }
+    }
+
+    return null;
+}
+
+fn hasTrait(trait_type: smithy.TraitType, traits: []smithy.Trait) bool {
+    return findTrait(trait_type, traits) != null;
+}
+
+const JsonMember = struct {
+    field_name: []const u8,
+    json_key: []const u8,
+    target: []const u8,
+    type_member: smithy.TypeMember,
+    shape_info: smithy.ShapeInfo,
+};
+
+fn getJsonMembers(allocator: std.mem.Allocator, shape: Shape, state: GenerationState) !?std.ArrayListUnmanaged(JsonMember) {
```

Comment:
There's about 400 LoC here dedicated to json serialization. Some types will need to serialize to XML, and CBOR is coming. I think it would be best to move the json serialization code to a separate file to establish a pattern here for the other serialization formats.

Reply:
I reorganised a bunch 4a6c84e